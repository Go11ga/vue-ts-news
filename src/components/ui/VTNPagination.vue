<template>
    <div v-if="totalCount !== 0" class="vtn-pagination">
        <VTNPaginationItem
            class="vtn-pagination__prev"
            label="Назад"
            :value="currentPage > 1 ? currentPage - 1 : 1"
        />

        <VTNPaginationItem
            v-for="(page, index) in pages"
            :key="index"
            :label="page.label"
            :value="page.value"
            :active="page.value == currentPage"
        />

        <VTNPaginationItem
            class="vtn-pagination__next"
            label="Вперёд"
            :value="currentPage < totalCount ? currentPage + 1 : totalCount"
        />
    </div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'
import VTNPaginationItem from '@/components/ui/VTNPaginationItem.vue'
import getPaginationOptions from '@/utils/paginationOptions'
import { PaginationItemOption } from '@/types/interfaces'

export default defineComponent({
    name: 'VTNPagination',

    props: {
        currentPage: {
            type: Number,
            required: true
        },

        totalCount: {
            type: Number,
            required: true
        }
    },

    components: {
        VTNPaginationItem
    },

    computed: {
        pages(): Array<PaginationItemOption> {
            const obj = getPaginationOptions({
                totalPages: this.totalCount,
                currentPage: this.currentPage
            })
            return obj
        }
    }
})
</script>
