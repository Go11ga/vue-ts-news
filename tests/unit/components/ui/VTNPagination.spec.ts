import { shallowMount } from '@vue/test-utils'
import VTNPagination from '@/components/ui/VTNPagination.vue'

describe('VTNPagination', () => {
    let wrapper = shallowMount(VTNPagination, {
        props: {
            currentPage: 1,
            totalCount: 10
        }
    })

    it('dummy name test', () => {
        expect(wrapper.vm.$options.name).toMatch('VTNPagination')
    })

    it('snapshot', () => {
        expect(wrapper.element).toMatchSnapshot()
    })

    it('отрисовка и пропсы', async () => {
        expect(wrapper.html().includes('vtn-pagination')).toBe(true)
        expect(wrapper.findAll('v-t-n-pagination-item-stub').length).toBe(12)
        expect(wrapper.findAll('v-t-n-pagination-item-stub')[0].classes()).toContain('vtn-pagination__prev')
        expect(wrapper.findAll('v-t-n-pagination-item-stub')[1].attributes().active).toBe("true")
        expect(wrapper.findAll('v-t-n-pagination-item-stub')[11].classes()).toContain('vtn-pagination__next')

        wrapper.setProps({
            currentPage: 3,
            totalCount: 5
        })

        await wrapper.vm.$nextTick()

        expect(wrapper.html().includes('vtn-pagination')).toBe(true)
        expect(wrapper.findAll('v-t-n-pagination-item-stub').length).toBe(7)
        expect(wrapper.findAll('v-t-n-pagination-item-stub')[0].classes()).toContain('vtn-pagination__prev')
        expect(wrapper.findAll('v-t-n-pagination-item-stub')[3].attributes().active).toBe("true")
        expect(wrapper.findAll('v-t-n-pagination-item-stub')[6].classes()).toContain('vtn-pagination__next')

        wrapper.setProps({
            currentPage: 3,
            totalCount: 1
        })

        await wrapper.vm.$nextTick()

        expect(wrapper.html().includes('vtn-pagination')).toBe(true)
        expect(wrapper.findAll('v-t-n-pagination-item-stub').length).toBe(3)
        expect(wrapper.findAll('v-t-n-pagination-item-stub')[0].classes()).toContain('vtn-pagination__prev')
        expect(wrapper.findAll('v-t-n-pagination-item-stub')[1].attributes().active).toBe("false")
        expect(wrapper.findAll('v-t-n-pagination-item-stub')[2].classes()).toContain('vtn-pagination__next')

        wrapper.setProps({
            currentPage: 3,
            totalCount: 0
        })

        await wrapper.vm.$nextTick()

        expect(wrapper.html().includes('vtn-pagination')).toBe(false)
    })

    it('computed pages', () => {
        let localThis = {
            paginationLength: 3,
            currentPage: 1,
            totalCount: 3
        }

        const component: any = VTNPagination

        expect(component.computed.pages.call(localThis)).toStrictEqual([
            { label: '1', value: 1 },
            { label: '2', value: 2 },
            { label: '3', value: 3 }
        ])
    })
})