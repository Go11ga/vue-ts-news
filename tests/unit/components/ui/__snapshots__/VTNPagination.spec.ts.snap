// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VTNPagination snapshot 1`] = `
<div
  class="vtn-pagination"
>
  <v-t-n-pagination-item-stub
    active="false"
    class="vtn-pagination__prev"
    label="Назад"
    value="1"
  />
  
  <v-t-n-pagination-item-stub
    active="true"
    label="1"
    value="1"
  />
  <v-t-n-pagination-item-stub
    active="false"
    label="2"
    value="2"
  />
  <v-t-n-pagination-item-stub
    active="false"
    label="3"
    value="3"
  />
  <v-t-n-pagination-item-stub
    active="false"
    label="4"
    value="4"
  />
  <v-t-n-pagination-item-stub
    active="false"
    label="5"
    value="5"
  />
  <v-t-n-pagination-item-stub
    active="false"
    label="6"
    value="6"
  />
  <v-t-n-pagination-item-stub
    active="false"
    label="7"
    value="7"
  />
  <v-t-n-pagination-item-stub
    active="false"
    label="8"
    value="8"
  />
  <v-t-n-pagination-item-stub
    active="false"
    label="9"
    value="9"
  />
  <v-t-n-pagination-item-stub
    active="false"
    label="10"
    value="10"
  />
  
  <v-t-n-pagination-item-stub
    active="false"
    class="vtn-pagination__next"
    label="Вперёд"
    value="2"
  />
</div>
`;
